packages/world/ts/actions/delegation.ts

```diff
@@ -0,0 +1,56 @@
+import type { WalletClient, Transport, Chain, Account, Hex, WalletActions, WriteContractReturnType } from "viem";
+import { getAction, getAbiItem, formatAbiItem, toFunctionSelector, type GetAbiItemParameters } from "viem/utils";
+import { writeContract } from "viem/actions";
+import { encodeSystemCallFrom, type SystemCallFrom } from "../encodeSystemCallFrom";
+
+type DelegationParameters = {
+  worldAddress: Hex;
+  delegatorAddress: Hex;
+  getSystemId: (functionSelector: Hex) => Hex;
+};
+
+// By extending clients with this function after delegation, the delegation automatically applies
+// to the World contract writes, meaning these calls are made on behalf of the delegator.
+export function delegation<TChain extends Chain, TAccount extends Account>({
+  worldAddress,
+  delegatorAddress,
+  getSystemId,
+}: DelegationParameters): (
+  client: WalletClient<Transport, TChain, TAccount>,
+) => Pick<WalletActions<TChain, TAccount>, "writeContract"> {
+  return (client) => ({
+    // Applies to: `client.writeContract`, `getContract(client, ...).write`
+    writeContract: (originalArgs): Promise<WriteContractReturnType> => {
+      // Skip if the contract isn't the World.
+      if (originalArgs.address !== worldAddress) {
+        return getAction(client, writeContract, "writeContract")(originalArgs);
+      }
+
+      // Ensured not to be `undefined` because of the `writeContract` args type.
+      const functionAbiItem = getAbiItem({
+        abi: originalArgs.abi,
+        name: originalArgs.functionName,
+        args: originalArgs.args,
+      } as unknown as GetAbiItemParameters)!;
+
+      // `callFrom` requires `systemId`.
+      const functionSelector = toFunctionSelector(formatAbiItem(functionAbiItem));
+      const systemId = getSystemId(functionSelector);
+
+      // Construct args for `callFrom`.
+      const callFromArgs: typeof originalArgs = {
+        ...originalArgs,
+        functionName: "callFrom",
+        args: encodeSystemCallFrom({
+          abi: originalArgs.abi,
+          from: delegatorAddress,
+          systemId,
+          functionName: originalArgs.functionName,
+          args: originalArgs.args,
+        } as unknown as SystemCallFrom),
```

Comment:
ooc why do we need to cast this?

Reply:
When wrapping this kind of viem functions that take an abi, we often need this kind of type cast. This is because it requires the actual abi types to provide strong typing. (in this case, to type functionName and args)

E.g., https://github.com/wevm/viem/blob/viem%402.7.19/src/actions/wallet/writeContract.ts#L156-L162